Misc CI bumps; replace cirrus freebsd with GHA; add all-done job

kolyshkin opened the pull request:
~~_Note: this currently includes #419 and #421 and is therefore a draft (will rebase once those are merged)._~~

This bumps various CI bits and pieces to their latest versions. See individual commits for details.

Also, since Cirrus CI is to be gone soon, let's replace it with GHA.

NOTE Once (or just before) merging this, you need to modify this repo branch protection rules:
 - require "all-done";
 - remove all other requirements.

kolyshkin committed:
ci: bump actions/checkout, actions/setup-go

... to their latest versions.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
ci: use v2 for actions-codespell

Using @master is more prone to some breakage.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
ci: add stable and oldstabe go to the matrix, drop old ones

Use "oldstable" (currently 1.25) and "stable" (currently 1.26) aliases,
so we always test the supported Go versions.

Drop the 1.21, 1.22, and 1.23 from the matrix as they are no longer
maintained / supported.

Finally, add a hack to test against the Go version specified in go.mod,
so we always test against a minimally supported version. For this to
work, we had to checkout sources before installing Go.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
ci: add all-done job

This job depends on every other one, and thus makes it easier to set up
"required" jobs in GHA CI (once this is merged, we only need "all-done"
as required).

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented:
@guelfey PTAL :pray:

Among the other things, will allow to have a single "required" job for this repo (the "all-done" one).

kolyshkin commented:
@guelfey PTAL

kolyshkin commented:
At this time I'd like to add Go 1.26 (and update other bits to their latest versions) but I'd like to receive some feedback first.

guelfey commented on .github/workflows/test.yml:
```diff
@@ -51,5 +51,14 @@ jobs:
+    - uses: codespell-project/actions-codespell@v2
+
+  all-done:
+    runs-on: ububtu-24.04
```
typo

kolyshkin replied:
fixed

guelfey commented on .github/workflows/test.yml:
```diff
@@ -27,17 +27,17 @@ jobs:
       fail-fast: false
       matrix:
-        go-version: ['1.20', '1.21', '1.22', '1.23']
+        go-version: ['1.20', oldstable, stable]
```
This means we don't explicitly need to bump to 1.26 though as this is now `stable`? I like it - I see the risk of the build breaking when a new Go version is released very low (but if it indeed happens, I want to know early). And keeping 1.20 in there also lets us know if we for some reason accidentally increase the minimum version requirement.

kolyshkin replied:
Yes, this means currently it's 1.25 and 1.26, and soon after the 1.27 release it will be changed to 1.26 and 1.27 with no input. I guess for a repo like this it makes sense.

The min version should be synced to one in go.mod. Actually I just went ahead and added a hack so that we don't have to specify the min version explicitly; please see the updated commit.

kolyshkin commented:
@guelfey I think I've addressed all the comments. Once (or just before) merging this, you need to modify repo branch protection rules to require "all-done" (and remove all the existing GHA checks).

kolyshkin committed:
ci: migrate freebsd tests from Cirrus to GHA

Cirrus CI is being shut down. Let's use lima to bring up FreeBSD VM and
test inside it.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented:
OK, I've also replaced `.cirrus.yml` (which ran freebsd tests) with an equivalent GHA job (running inside a lima VM), because, unfortunately, Cirrus CI will cease to exist soon.
